Fix #2805 and #2398: Nested JavaScript classes.

sjrd opened the pull request:
This commit dramatically changes the way nested JS classes are compiled, to fix a pervasive issue about their semantics. We call a class "nested" if it either *inner* (defined in the body of a class, trait, or non-static object) or *local* (defined inside a `def`).

## Problem with the old semantics

The semantic issue is that we have always allowed the definition of nested non-native JS classes, but using the normal compilation scheme of scalac, driven by explicitouter and lambdalift. This caused their constructor to be visible to JavaScript (and JS interop within Scala.js) as having more arguments, for the captures. For Scala classes, this is not an issue, as their constructor cannot be manipulated by JavaScript. But for JavaScript classes, it is probablematic.

For example, consider:
```scala
class Enclosing(val x: Int) {
  class InnerJSClass extends js.Object {
    def getX(): Int = x
  }
}
```
Because of `explicitouter`, the constructor of `InnerJSClass` takes an additional parameter `$outer` of type `Enclosing`. This means that if we ask for `js.constructor[enclosing.InnerJSClass]`, we receive a class value that is basically unusable.

For local classes, lambdalift would add zero to many parameters for captures of local vals and vars.

For nested *native* JS classes, there was "no issue" in the sense that they were disallowed as compile errors to begin with. However, this is a limitation, as evidenced by the issue #2398.

## New semantics

The new semantics for nested JS classes are the "obvious" ones. If they capture stuff (the outer pointer and/or local vals), they are properly *captured* by the JS class value, and are therefore not visible as constructor parameters.

Additionally, and relatedly, this means that for inner JS classes, there is a different JS class value for each enclosing instance. For local JS classes, there is a different JS class value for each time the code "executes" the class declaration.

Together, these changes to the semantics address #2805.

Given all the infrastructure that is set up for the above, it also becomes trivial (and more consistent) to address #2398 in the same sweep. Indeed, it is virtually only a matter of lifting compile-time checks.

## Implementation

Implementing the above semantics is far from being trivial, because erasure tends to destroy a lot of information that we need. In particular, it will destroy the *paths* of inner JS classes (or *prefixes* in scalac terms). flatten will also destroy the identity of nested JS classes, preventing from distinguishing inner classes by their enclosing instance and local classes by their "execution path".

We address this with actual code transformations of the scalac trees, something that has no real precedent in our compiler plugin. We actually reify nested JS classes, in two steps, before erasure kicks in and destroys what we need.

The first step, in phase `xplicitinnerjs`, creates for every inner JS class, a field in its enclosing class to hold its JS class value. The second step, in phase `xplicitlocaljs`, does something similar with local vals for local JS classes, and transforms all references to preserve identities. Ideally, this would be done in a single phase, but we need to do some work *before* the phase `fields` of Scala 2.12+, and some *after* that phase. In particular, we need to

* create the inner JS class fields before `fields`, so that the latter phase will correctly mix them in subclasses of traits, and
* desugar references to module classes after `fields`, since it is that phase that reifies modules.

Since there can be dependencies of all sorts between *references* to inner- and local JS classes, all references must be processed in the same phase. This is why `xplicitlocaljs` actually desugars all references, including those to inner JS classes. `xplicitinnerjs` only creates the fields that will hold the JS class values.

For further information on the implementation, and code snippets before and after the transformations, refer to the long Scaladoc comments of those two phases.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4316/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4316/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5797/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5797/</a>
Test FAILed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4318/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4318/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5799/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5799/</a>
Test FAILed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4319/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4319/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5800/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5800/</a>
Test FAILed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4320/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4320/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5801/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5801/</a>
Test FAILed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4321/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4321/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5803/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5803/</a>
Test PASSed.

sjrd commented:
:tada:

Now to write more tests for the new features :P

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4322/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4322/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5804/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5804/</a>
Test FAILed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4323/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4323/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5806/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5806/</a>
Test PASSed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4324/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4324/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5807/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5807/</a>
Test PASSed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4326/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4326/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5810/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5810/</a>
Test PASSed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4327/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4327/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5812/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5812/</a>
Test PASSed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4328/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4328/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5813/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5813/</a>
Test PASSed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4329/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4329/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5814/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5814/</a>
Test PASSed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4331/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4331/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5818/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5818/</a>
Test PASSed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4332/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4332/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5819/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5819/</a>
Test PASSed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4333/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4333/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5822/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5822/</a>
Test PASSed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4334/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4334/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5824/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5824/</a>
Test PASSed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4335/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4335/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5826/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5826/</a>
Test PASSed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4336/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4336/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5827/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5827/</a>
Test PASSed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4337/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4337/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5828/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5828/</a>
Test PASSed.

sjrd commented:
Aaand, it is complete at last! :smile: 

@gzm0 Have fun reviewing this one :stuck_out_tongue_closed_eyes:

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4338/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4338/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5829/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5829/</a>
Test PASSed.

sjrd commented on partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.0/neg/t6446-additional.check:
```diff
@@ -11,21 +11,23 @@ superaccessors   8  add super accessors in traits and nested classes
     extmethods   9  add extension methods for inline classes
```
[Conveniently jump forward to the last updated .check files to skip the noise](#diff-91719e27cf3802a8f2167e22dbea938d)

sjrd commented on partest-suite/src/test/resources/scala/tools/partest/scalajs/2.13.0-M2/neg/t7494-no-options.check:
```diff
@@ -12,20 +12,22 @@ superaccessors   8  add super accessors in traits and nested classes
+    delambdafy  27  remove lambdas
+           jvm  28  generate JVM bytecode
+       ploogin  29  A sample phase that does so many things it's kind of hard...
+      terminal  30  the last phase during a compilation run
```
[Conveniently jump backward to the first updated .check files to skip the noise](#diff-a2268a717bb0a4cd30d9c184640ef6f2)

sjrd commented:
Rebased.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4373/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4373/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5882/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5882/</a>
Test PASSed.

gzm0 commented on test-suite/js/src/test/scala/org/scalajs/testsuite/jsinterop/NestedJSClassTest.scala:
```diff
@@ -0,0 +1,596 @@
+    assertEquals("salutworld2helloworld2", inner2.foobar())
+
+    assertTrue((inner2: Any).isInstanceOf[container2.InnerJSSubclass])
+    assertFalse((inner2: Any).isInstanceOf[container1.InnerJSSubclass])
```
Why does this need an ascription to `Any`?

sjrd replied:
To avoid the compiler warning "inner2 with type `js.Dynamic` can never be an instance of `container2.InnerJSSubclass`"

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/ClassEmitter.scala:
```diff
@@ -70,40 +70,71 @@ private[emitter] final class ClassEmitter(jsGen: JSGen) {
-          envFieldDef("b", className, js.Undefined(), mutable = true)
+        val genStoreJSSuperClass = tree.jsSuperClass.map { jsSuperClass =>
+          for (rhs <- desugarExpr(jsSuperClass, resultType = AnyType)) yield {
+            js.VarDef(genStoredSuperClass().ident, Some(rhs))
```
Not sure `genStoredSuperClass` deserves to be special. Just use the `envField` methods?

sjrd replied:
Like you said in the other comment, it's indeed not an abuse of the mechanism (just of the name). So indeed it doesn't deserve to be special.

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/FunctionEmitter.scala:
```diff
@@ -2284,8 +2313,10 @@ private[emitter] class FunctionEmitter(jsGen: JSGen) {
+          if (env.isLocalVar(name))
+            js.VarRef(transformLocalVarIdent(name))
+          else
+            envField("cc", name.name, name.originalName)
```
We should reconsider the name `envField`: `cc` and `superClass` are both fair usages of the `envField` "infrastructure" but blatant abuses of the (implied) semantics. We should name it something like `internalVariable` (but shorter).

This should probably happen later, not in this PR.

sjrd replied:
True. Also I think `b` has the same kind of usage.

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/FunctionEmitter.scala:
```diff
@@ -2474,7 +2517,12 @@ private object FunctionEmitter {
+      vars.getOrElse(ident.name, {
+        // It must be a JS class capture, which must be immutable
+        false
+      })
```
This syntax is quite odd.

Have you considered:
```scala
// If we do not know the var, it must be a JS class capture, which must be immutable
vars.getOrElse(ident.name, false)
```

sjrd replied:
Doesn't seem odd to me given that `getOrElse` takes a by-name argument for `default`. But I can change it.

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/linker/checker/IRChecker.scala:
```diff
@@ -1238,7 +1307,8 @@ private final class IRChecker(unit: LinkingUnit,
             reportDuplicateLocalVarName(name)
```
I see this crossed my PR for name uniqueness. Do you think we can unify this now? It's a bit awkward to have almost the same loop twice.

sjrd replied:
I will investigate.

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala:
```diff
@@ -4039,6 +4084,47 @@ abstract class GenJSCode extends plugins.PluginComponent
+          val captureValues = {
+            if (code == CREATE_INNER_JS_CLASS) {
+              val outer = genThis()
+              List.fill(classSym.info.decls.count(_.isClassConstructor))(outer)
```
Doesn't this make adding a constructor to a public nested class a binary incompatible change? That would be a problem, no?

sjrd replied:
No, it doesn't, because calls to `createInnerJSClass` are always in the same compilation unit as the inner class itself, so they will always be in sync. If you add a constructor to that inner class, anyway you're recompiling the enclosing class, which will therefore receive a new call `createInnerJSClass`.

sjrd commented on ir/src/main/scala/org/scalajs/core/ir/Trees.scala:
```diff
@@ -858,13 +858,31 @@ object Trees {
   final class ClassDef(
       val name: Ident,
       val kind: ClassKind,
+      val jsClassCaptures: Option[List[ParamDef]],
```
Yes, I'll add a comment.
* `jsClassCaptures = None` means that the class is top-level. There is only one in the world, and its unique class value can be loaded with a `LoadJSConstructor`. `CreateJSClass` is invalid for such a `ClassDef`.
* `jsClassCaptures = Some(Nil)` means that the class not top-level, so `LoadJSConstructor` is not valid. New class values for that `ClassDef` can be created with `CreateJSClass`. It just so happens that there is no capture, so all class values created from that `ClassDef` will be "identical" (although they won't be `eq`).

sjrd commented on ir/src/main/scala/org/scalajs/core/ir/Trees.scala:
```diff
@@ -858,13 +858,31 @@ object Trees {
+      val jsClassCaptures: Option[List[ParamDef]],
       val superClass: Option[Ident],
       val interfaces: List[Ident],
+      val jsSuperClass: Option[Tree],
```
Not purely informational. It is still used for all metadata-like things. Notably:
* It determines the subtyping relationships for arrays, i.e., whether `Array[Foo] <: Array[Bar]` depends on `Foo`'s `superClass` and `interfaces`, even when `Foo` and `Bar` are JS types, and whether or not they have class captures or a `jsSuperClass.isDefined`.
* It determines the value of `classOf[Foo].getSuperclass()`.

sjrd commented on ir/src/test/scala/org/scalajs/core/ir/PrintersTest.scala:
```diff
@@ -855,7 +855,9 @@ class PrintersTest {
     import ClassKind._
 
```
Ah, indeed!

sjrd commented on test-suite/js/src/test/scala/org/scalajs/testsuite/compiler/InteroperabilityTest.scala:
```diff
@@ -182,6 +182,59 @@ class InteroperabilityTest {
 
+  @Test def nestedNativeJSClasses(): Unit = {
+    js.eval("""
+      var InteroperabilityTestContainerClass = (function(a) {
```
Hum, could you elaborate? I don't understand what you mean by "we don't copy it".

gzm0 replied:
I meant to make sure we don't copy the outer class instead of having a reference to it. But on second thought that's probably overkill. Seems more of a C++ like problem to have :)

sjrd commented on compiler/src/main/scala/org/scalajs/core/compiler/ExplicitLocalJS.scala:
```diff
@@ -0,0 +1,412 @@
+ *  - If `C` is a statically accessible class, desugar to
+ *    `runtime.constructorOf(classOf[C])` so that the reified symbol survives
+ *    erasure and reaches the back-end.
+ *  - If `C` is an inner JS class, it must be of the form `path.D`, and we
```
No here `C := path.D`. If I said `C` is of the form `path.C`, that wouldn't make sense as I would be recursively defining `C := path.C`. So `D`, just like `path`, is whatever fits so that `path.D` gives `C`. In a sense, I'm pattern-matching over `C`, and this is the `case path.C =>` where both `path` and `C` are "captures" from the pattern match.

Maybe it would clearer to add "it must be of the form `path.D` for some pair (`path`, `D`)"?

sjrd commented on compiler/src/main/scala/org/scalajs/core/compiler/ExplicitLocalJS.scala:
```diff
@@ -0,0 +1,412 @@
+  }
+
+  private def jsclassAccessorFor(clazz: Symbol): Symbol =
+    clazz.owner.info.member(clazz.name.append("$jsclass").toTermName)
```
None, they should indeed be shared.

sjrd commented on compiler/src/main/scala/org/scalajs/core/compiler/ExplicitLocalJS.scala:
```diff
@@ -0,0 +1,412 @@
+          gen.mkAttributedCast(genJSConstructorOf(tree, newTpeArg.tpe),
+              JSDynamicClass.tpe)
+
+        // Translate x.isInstanceOf[T] for inner and local JS classes
```
That's right.

sjrd commented on compiler/src/main/scala/org/scalajs/core/compiler/ExplicitLocalJS.scala:
```diff
@@ -0,0 +1,412 @@
+          val newTpeArg = transform(tpeArg)
+          atPos(tree.pos) {
+            localTyper.typed {
+              val jsCtorOf = genJSConstructorOf(tree, newTpeArg.tpe)
```
It doesn't matter. `typed { ... }` does not take a by-name argument, it's a normal argument (same with `atPos`, btw). I can pull it out if you find that less confusing.

sjrd commented on compiler/src/main/scala/org/scalajs/core/compiler/ExplicitLocalJS.scala:
```diff
@@ -0,0 +1,412 @@
+
+      if (clazz.isTrait || clazz.isModuleClass) {
+        throw new TypeError(tree.pos,
+            s"non-trait class type required but $tpe found")
```
Because it should have been checked earlier in PrepJSInterop. If we arrive here and we fall in that case, it's a compiler bug (or a bug of another compiler plugin trying to mess with us).

sjrd commented on compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala:
```diff
@@ -558,10 +564,15 @@ abstract class GenJSCode extends plugins.PluginComponent
+      val (jsClassCaptures, generatedConstructor) =
+        genJSClassCapturesAndConstructor(sym, constructorTrees.toList)
+
+      val jsSuperClass = jsClassCaptures.map(_.head.ref)
```
It doesn't. It has other arguments for all the other captures of the class (including the `outer` params). But `GenJSCode` arranges things so that the capture of the super class always comes first.

Note that this is not a requirement of the IR. In fact in the IR nothing says you need a capture for your superclass (`ClassDef.jsSuperClass` could refer exclusively to top-level stuff). But our compiler will currently always generate things that way.

sjrd commented on compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala:
```diff
@@ -1036,27 +1053,38 @@ abstract class GenJSCode extends plugins.PluginComponent
+        (None, ctorDef)
       } else {
-        withNewLocalNameScope {
+        withNewLocalNameScopeUsingJSSuperClassParamName {
```
The difference between the two is that:
* `JSSuperClassParamName` is a capture param, referencing the superclass value of the enclosing `ClassDef`. It is only available in *instance members* (not static members) of the `ClassDef`, and therefore cannot be used in regular `genMethodBody`'s, because the `DefDef`s in a non-native JS class are compiled as static methods taking `this` as a parameter. It is only used for the dispatchers in `GenJSExports`, which are always used for instance members (methods and properties) and never for static methods.
* `jsSuperClassValueForSuperCall` is a mechanism to ensure that the static methods generated for `DefDef`s somehow receive a reference to the superclass value. Since they cannot use `JSSuperClassParamName`, they need to receive it in some other way. We make sure they get it through `lambdalift` doing its job, by explicitly accessing that superclass value as a term inside `innerOrLocalJSSuperCall`.

We cannot use a scoped variable because, being a class-level capture, it must be exactly the same in the entire class. But we generate a different local-name scope for each dispatcher method/property (as we should), so we need an external way to enforce that all the local-name scopes agree on the name of that class-level capture; they do not have the authority to choose it.

sjrd commented on compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala:
```diff
@@ -2398,7 +2438,12 @@ abstract class GenJSCode extends plugins.PluginComponent
+              jsSuperClassValueForSuperCall := None
+          ) {
+            genJSSuperCall(tree, explicitJSSuperClassValue, isStat)
+          }
```
I think there was a reason at some point in the development, but now it seems there's no reason not to, indeed.

sjrd commented on compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala:
```diff
@@ -4039,6 +4084,47 @@ abstract class GenJSCode extends plugins.PluginComponent
+        val classSym = resolveReifiedJSClassSym(args(0))
+        val superClassValue = genExpr(args(1))
+        if (classSym == NoSymbol) {
+          js.Undefined()
```
Because `resolveReifiedJSClassSym` generates an error already.

sjrd commented on compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala:
```diff
@@ -4039,6 +4084,47 @@ abstract class GenJSCode extends plugins.PluginComponent
+        val jsSuperClassValue = genExpr(args(0))
+        val newTree = args(1)
+        assert(newTree.isInstanceOf[Apply], pos)
+        genPrimitiveJSNew(newTree.asInstanceOf[Apply], Some(jsSuperClassValue))
```
It's a bit weird. It's because the semantics of the class value are not the same: in the JS_CLASS case, it refers to the class value of that very class, but in the JS_OBJECT, it refers to the class value *of the super class*.

I was also thinking of merging both of them with `INNER_LOCAL_JS_SUPER_CALL`, and simply keep a unique `WITH_CONTEXTUAL_JS_CLASS_VALUE` primitive that would store the class value in `contextualJSCLassValue` (replacing `jsSuperClassValueForSuperCall`).

sjrd commented on test-suite/js/src/test/scala/org/scalajs/testsuite/jsinterop/NestedJSClassTest.scala:
```diff
@@ -0,0 +1,596 @@
+      js.constructorOf[InnerJSClass]
+  }
+
+}
```
Because that's not allowed at compile-time. See
* https://github.com/sjrd/scala-js/blob/f82de5f73c49bec86bc1d4090403e559e1accf93/compiler/src/test/scala/org/scalajs/core/compiler/test/JSInteropTest.scala#L778
* https://github.com/sjrd/scala-js/blob/f82de5f73c49bec86bc1d4090403e559e1accf93/compiler/src/test/scala/org/scalajs/core/compiler/test/JSInteropTest.scala#L1268

That said I think I have somehow removed the test forbidding a native JS class/trait/object inside a non-native JS class/trait. In current master it's done in `JSInteropTest.notNested`, but that test was somehow restricted to `notNestedInScalaClassTrait`. There should also be `notNestedInNonNativeJSClassTrait`.

sjrd commented:
Updated. Unless I missed something, I should have addressed all your comments except https://github.com/scala-js/scala-js/pull/3161#discussion_r151839695. I kept changes as separate commits for easier review, but obviously they should be squashed before merging.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4385/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4385/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5900/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5900/</a>
Test FAILed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4386/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4386/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5901/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5901/</a>
Test PASSed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4387/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4387/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5903/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5903/</a>
Test PASSed.

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/JSGlobalAddons.scala:
```diff
@@ -45,6 +45,50 @@ trait JSGlobalAddons extends JSDefinitions
+   *  path.D[?0, ?1] forSome { type ?0; type ?1 <: List[Seq[?0]] }
+   *  }}}
+   */
+  def reinventTypeParams(tp: Type): Type = {
```
Is this used outside of `extractSuperTpeFromImpl`? If no, make private?

sjrd replied:
It used to be. Not anymore, though, so I'll make it private :)

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/JSEncoding.scala:
```diff
@@ -40,18 +40,43 @@ trait JSEncoding extends SubComponent { self: GenJSCode =>
+      usedLocalNames += JSSuperClassParamName
+      body
+    }
+  }
```
Consider replacing this method with a `reserveName` method that is only allowed to be called if `localSymbolNames` is empty.

sjrd commented on tools/shared/src/main/scala/org/scalajs/core/tools/linker/checker/IRChecker.scala:
```diff
@@ -86,6 +89,43 @@ private final class IRChecker(unit: LinkingUnit,
+      }
+
+      classCaptures.foldLeft(Set.empty[String]) {
+        case (alreadyDeclared, p @ ParamDef(ident, tpe, mutable, rest)) =>
```
No, it's not the same scope. Class captures are one scope above. So yes, params and locals are allowed to shadow class captures.

(We actually need this to be valid, or we'll need much more infrastructure in `GenJSCode`.)

sjrd commented:
Updated.

gzm0 commented:
LGTM

sjrd commented:
Squashed and rebased :)

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4401/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4401/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5922/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5922/</a>
Test PASSed.

sjrd commented:
Do not merge until #3204 is fast-forwarded to master.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4403/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4403/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5924/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5924/</a>
Test PASSed.
